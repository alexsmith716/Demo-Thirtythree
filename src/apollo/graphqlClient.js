import fetch from 'isomorphic-fetch';

const graphqlClient = ({ endpoint, query, variables={}, method='POST' }) => {

	//	console.log('>>>> graphqlClient > endpoint: ', endpoint);
	//	console.log('>>>> graphqlClient > query: ', query);
	//	console.log('>>>> graphqlClient > variables: ', {...variables});
	//	console.log('>>>> graphqlClient > method: ', method);

	return fetch(endpoint, {
		method: method,
		headers: {
			'Content-Type': 'application/json',
		},
		body: JSON.stringify({
			query: query,
			variables: {...variables}
		}),
	})
		.then(response => response.json())
		.then(data => {
			if (data.error) {
				throw new Error(data.error);
			}
			console.log('>>>> graphqlClient > DATA:', JSON.stringify(data));
			return data;
		})
		.catch(err => {
			console.error('>>>> graphqlClient > ERROR:', err.message);
			return null;
		});
};

export default graphqlClient;
