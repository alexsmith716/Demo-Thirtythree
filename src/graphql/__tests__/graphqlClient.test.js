import nock from 'nock'

import graphqlClient from '../../apollo/graphqlClient';

const mockRickAndMortyCharacter = {
	id: '11',
	name: 'Albert Einstein',
	status: 'Dead',
	species: 'Human',
	type: '',
	gender: 'Male',
	origin: {
		name: 'Earth (C-137)',
		type: 'Planet',
		dimension: 'Dimension C-137'
	},
	location: {
		name: 'Earth (Replacement Dimension)',
		type: 'Planet',
		dimension: 'Replacement Dimension'
	},
	image: 'https://rickandmortyapi.com/api/character/avatar/11.jpeg',
	episode: [ { name: 'A Rickle in Time', episode: 'S02E01' } ]
}

const GET_RICK_AND_MORTY_CHARACTER = `
	query GetRickAndMortyCharacter($id: ID!) {
		character(id: $id) {
			id
			name
			status
			species
			type
			gender
			origin {
				name
				type
				dimension
			}
			location {
				name
				type
				dimension
			}
			image
			episode {
				name
				episode
			}
		}
	}
`;

describe('GraphqlClient', () => {

	describe('graphqlClient', () => {
		afterEach(() => {
			nock.cleanAll()
		});

		it('checks endpoint fetch attempted', async () => {
			nock('https://rickandmortyapi.com').post('/graphql').reply(200, mockRickAndMortyCharacter);

			const res = await graphqlClient({endpoint: 'https://rickandmortyapi.com/graphql', query: GET_RICK_AND_MORTY_CHARACTER, variables: {id: 11}});
			expect(res).toMatchSnapshot();
		})

	})
});
