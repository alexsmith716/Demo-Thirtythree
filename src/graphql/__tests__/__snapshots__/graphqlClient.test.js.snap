// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphqlClient graphqlClient checks endpoint fetch attempted 1`] = `
Object {
  "episode": Array [
    Object {
      "episode": "S02E01",
      "name": "A Rickle in Time",
    },
  ],
  "gender": "Male",
  "id": "11",
  "image": "https://rickandmortyapi.com/api/character/avatar/11.jpeg",
  "location": Object {
    "dimension": "Replacement Dimension",
    "name": "Earth (Replacement Dimension)",
    "type": "Planet",
  },
  "name": "Albert Einstein",
  "origin": Object {
    "dimension": "Dimension C-137",
    "name": "Earth (C-137)",
    "type": "Planet",
  },
  "species": "Human",
  "status": "Dead",
  "type": "",
}
`;
